packages/core-commands/src/site-editor-navigation-commands.js

```diff
@@ -48,30 +46,42 @@ const getNavigationCommandLoaderPerPostType = ( postType ) =>
 					'getEntityRecords',
 					[ 'postType', postType, query ]
 				),
-				canvasMode: unlock( select( editSiteStore ) ).getCanvasMode(),
+				// We're using the string literal to check whether we're in the site editor.
+				/* eslint-disable-next-line @wordpress/data-no-store-string-literals */
+				isSiteEditor: !! select( 'edit-site' ),
 			};
 		}, deps );
 
 		const commands = useMemo( () => {
 			return ( records ?? [] ).slice( 0, 10 ).map( ( record ) => {
+				const isSiteEditor = getPath( window.location.href )?.includes(
+					'site-editor.php'
+				);
+				const extraArgs = isSiteEditor
+					? { canvas: getQueryArg( window.location.href, 'canvas' ) }
+					: {};
 				return {
 					name: record.title?.rendered + ' ' + record.id,
 					label: record.title?.rendered
 						? record.title?.rendered
 						: __( '(no title)' ),
 					icon: icons[ postType ],
 					callback: ( { close } ) => {
-						history.push( {
+						const args = {
 							postType,
 							postId: record.id,
-							canvas:
-								canvasMode === 'edit' ? canvasMode : undefined,
-						} );
+							...extraArgs,
+						};
+						const targetUrl = addQueryArgs(
+							'site-editor.php',
+							args
+						);
+						document.location = targetUrl;
```

Comment:
Here I've tried using `window.history` but failed to trigger the site editor to reload its content. I'm not sure if there's a way to trigger "client site routing" without having access to the "history" object. Anyone knows?